dialogue-core/src/main/java/com/palantir/dialogue/core/ClientPool.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.palantir.dialogue.Channel;
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * Facilitates creating many clients which all share the same connection pool. Should only create one of these per
+ * server. Close it when your server shuts down to release resources.
+ */
+public interface ClientPool extends Closeable {
+
+    /** Returns an implementation of the given dialogueInterface, hooked up to a fresh smart channel underneath. */
+    <T> T get(Class<T> dialogueInterface, Listenable<DialogueConfig> config);
+
+    /**
+     * Returns a channel for interacting with the given abstract upstream service, which routes traffic
+     * appropriately to the various available nodes. Live-reloaded every time the urls change.
+     */
+    Channel smartChannel(Listenable<DialogueConfig> config);
+
+    /**
+     * Gets a direct channel to a single host within the specified Config. Live-reloads under the hood. The channel
+     * will always fail if the specified uri is not listed in the latest version of the config. Somewhat dangerous
+     * because this has no limits / failover.
+     */
+    Channel rawHttpChannel(String uri, Listenable<DialogueConfig> config);
```

Comment:
is this actually something that should be exposed?

Reply:
Probably not to be honest - I started with this 3-membre interface just to hash out in my head what the key building blocks were. Happy to push it down to package private.